File: lib/src/models/game_record.dart
class GameRecord {
  final String id;
  final String platform;
  final String username;
  final String opponent;
  final String pgn;
  final String? timeControl;
  final bool rated;
  final String? result;
  final DateTime? playedAt;
  final DateTime createdAt;
  final DateTime? analyzedAt;

  GameRecord({
    required this.id,
    required this.platform,
    required this.username,
    required this.opponent,
    required this.pgn,
    this.timeControl,
    this.rated = false,
    this.result,
    this.playedAt,
    required this.createdAt,
    this.analyzedAt,
  });

  factory GameRecord.fromJson(Map<String, dynamic> json) {
    return GameRecord(
      id: json['id'] as String,
      platform: json['platform'] as String,
      username: json['username'] as String,
      opponent: json['opponent'] as String,
      pgn: json['pgn'] as String,
      timeControl: json['time_control'] as String?,
      rated: json['rated'] as bool? ?? false,
      result: json['result'] as String?,
      playedAt: json['played_at'] != null
          ? DateTime.parse(json['played_at'] as String)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      analyzedAt: json['analyzed_at'] != null
          ? DateTime.parse(json['analyzed_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toInsertJson() {
    return {
      'platform': platform,
      'username': username,
      'opponent': opponent,
      'pgn': pgn,
      'time_control': timeControl,
      'rated': rated,
      'result': result,
      'played_at': playedAt?.toIso8601String(),
    };
  }
}

File: src/models/gameRecord.ts
export type GamePhase = 'opening' | 'middlegame' | 'endgame';

export interface GameRecord {
  id: string;
  platform: string;
  username: string;
  opponent: string;
  pgn: string;
  timeControl: string | null;
  rated: boolean;
  result: string | null;
  playedAt: Date | null;
  createdAt: Date;
  analyzedAt: Date | null;
  eco: string | null;
  openingName: string | null;
  userColor: 'white' | 'black' | null;
  userRating: number | null;
  opponentRating: number | null;
  clockPerPly: number[] | null;
  totalPlies: number | null;
  parsedMetadataAt: Date | null;
}

function parseColor(v: unknown): 'white' | 'black' | null {
  return v === 'white' || v === 'black' ? v : null;
}

function parseClockPerPly(v: unknown): number[] | null {
  if (!Array.isArray(v)) return null;
  const out: number[] = [];
  for (const item of v) {
    if (typeof item === 'number' && Number.isFinite(item)) out.push(item);
  }
  return out;
}

export function gameRecordFromJson(json: any): GameRecord {
  return {
    id: json.id as string,
    platform: json.platform as string,
    username: json.username as string,
    opponent: json.opponent as string,
    pgn: json.pgn as string,
    timeControl: (json.time_control as string | null) ?? null,
    rated: (json.rated as boolean | null) ?? false,
    result: (json.result as string | null) ?? null,
    playedAt: json.played_at ? new Date(json.played_at as string) : null,
    createdAt: new Date(json.created_at as string),
    analyzedAt: json.analyzed_at ? new Date(json.analyzed_at as string) : null,
    eco: (json.eco as string | null) ?? null,
    openingName: (json.opening_name as string | null) ?? null,
    userColor: parseColor(json.user_color),
    userRating: (json.user_rating as number | null) ?? null,
    opponentRating: (json.opponent_rating as number | null) ?? null,
    clockPerPly: parseClockPerPly(json.clock_per_ply),
    totalPlies: (json.total_plies as number | null) ?? null,
    parsedMetadataAt: json.parsed_metadata_at ? new Date(json.parsed_metadata_at as string) : null,
  };
}

/** ECO family (`B33` → `B3*`) used for opening grouping. */
export function ecoFamily(eco: string | null): string | null {
  if (!eco || eco.length < 2) return null;
  return `${eco[0].toUpperCase()}${eco[1]}*`;
}
